opentelemetry-etw-logs/src/logs/exporter.rs

```diff
@@ -0,0 +1,461 @@
+        field_tag: u32,
+    ) {
+        let event_time: SystemTime = log_data
+            .record
+            .timestamp
+            .unwrap_or(log_data.record.observed_timestamp);
+
+        const COUNT_TIME: u8 = 1u8;
+        const PART_A_COUNT: u8 = COUNT_TIME;
+        event.add_struct("PartA", PART_A_COUNT, field_tag);
+        {
+            let timestamp = win_filetime_from_systemtime!(event_time);
+            event.add_filetime("time", timestamp, tld::OutType::Default, field_tag);
+        }
+    }
+
+    fn populate_part_b(
+        &self,
+        event: &mut tld::EventBuilder,
+        log_data: &opentelemetry_sdk::export::logs::LogData,
+        level: tld::Level,
+        event_id: Option<i64>,
+        event_name: Option<&str>,
+    ) {
+        // Count fields in PartB
+        const COUNT_TYPE_NAME: u8 = 1u8;
+        const COUNT_SEVERITY_NUMBER: u8 = 1u8;
+
+        let field_count = COUNT_TYPE_NAME
+            + COUNT_SEVERITY_NUMBER
+            + log_data.record.body.is_some() as u8
+            + log_data.record.severity_text.is_some() as u8
+            + event_id.is_some() as u8
+            + event_name.is_some() as u8;
+
+        // Create PartB struct
+        event.add_struct("PartB", field_count, 0);
+
+        // Fill fields of PartB struct
+        event.add_str8("_typeName", "Logs", tld::OutType::Default, 0);
+
+        if let Some(body) = log_data.record.body.clone() {
+            add_attribute_to_event(event, &Key::new("body"), &body);
+        }
+
+        event.add_u8("severityNumber", level.as_int(), tld::OutType::Default, 0);
+
+        if let Some(severity_text) = &log_data.record.severity_text {
+            event.add_str8(
+                "severityText",
+                severity_text.as_ref(),
+                tld::OutType::Default,
+                0,
+            );
+        }
+
+        if let Some(event_id) = event_id {
+            event.add_i64("eventId", event_id, tld::OutType::Default, 0);
+        }
+
+        if let Some(event_name) = event_name {
+            event.add_str8("name", event_name, tld::OutType::Default, 0);
+        }
+    }
+
+    fn populate_part_c<'a>(
+        &'a self,
+        event: &mut tld::EventBuilder,
+        log_data: &'a opentelemetry_sdk::export::logs::LogData,
+        field_tag: u32,
+    ) -> (Option<i64>, Option<&str>) {
+        //populate CS PartC
+        let mut event_id: Option<i64> = None;
+        let mut event_name: Option<&str> = None;
+
+        if let Some(attr_list) = &log_data.record.attributes {
+            let mut cs_c_count = 0;
+
+            // find if we have PartC and its information
+            for (key, value) in attr_list.iter() {
+                match (key.as_str(), &value) {
+                    (EVENT_ID, AnyValue::Int(value)) => {
+                        event_id = Some(*value);
+                        continue;
+                    }
+                    (EVENT_NAME_PRIMARY, AnyValue::String(value)) => {
```

Comment:
i think we should remove this event primary secondary logic, and always populate evenname into partb only.

Reply:
I think this was to take care of a scenario where the user is using a version of tokio-tracing library without `name` metadata support. As we can't control what version of library they may be using. Unless we want to document that as a requirement, and then remove it from both user_events and ETW exporters.

Reply:
I tried to honour what user_events has. IMO, we may blend these two projects together at some point to avoid repeating code and decisions.